src/gpt.c

```diff
@@ -0,0 +1,346 @@
+
+	g_return_val_if_fail(cxt, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	grain_size = fdisk_get_grain_size(cxt);
+	sector_size = fdisk_get_sector_size(cxt);
+
+	if (fdisk_get_collision(cxt)) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Partition table collision found: %s",
+				fdisk_get_collision(cxt));
+		goto out;
+	}
+	if (!fdisk_is_labeltype(cxt, FDISK_DISKLABEL_GPT)) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"GPT not found");
+		goto out;
+	}
+	if (fdisk_gpt_is_hybrid(cxt)) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Hybrid GPT is not supported, use a protective MBR instead");
+		goto out;
+	}
+	if (fdisk_get_alignment_offset(cxt) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Non-zero aligment offset (%ld) is not supported",
+				fdisk_get_alignment_offset(cxt));
+		goto out;
+	}
+
+	if (grain_size < sector_size)
+		g_error("%s: libfdisk reported grain size is less than sector size\n", G_STRLOC);
+
+	res = TRUE;
+
+out:
+	return res;
+}
+
+static gboolean check_region(struct fdisk_context *cxt,
+		guint64 region_start, guint64 region_size,
+		GError **error)
+{
+	gboolean res = FALSE, found = FALSE;
+	guint64 grain_size, sector_size;
+	struct fdisk_table *tb = NULL;
+	struct fdisk_iter *itr = NULL;
+	struct fdisk_partition *pa = NULL;
+	fdisk_sector_t number_of_sectors;
+	fdisk_sector_t region_start_sector, region_end_sector;
+
+	g_return_val_if_fail(cxt, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	grain_size = fdisk_get_grain_size(cxt);
+	sector_size = fdisk_get_sector_size(cxt);
+
+	if (region_start < 34*512 || region_size == 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Region configuration is invalid");
+		goto out;
+	}
+
+	if ((region_start % grain_size) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Region start %"G_GINT64_MODIFIER "d is not aligned to grain-size %"G_GINT64_MODIFIER "d",
+				region_start, 2 * grain_size);
+		goto out;
+	}
+
+	if ((region_size % (2 * grain_size)) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Region size %"G_GINT64_MODIFIER "d is not aligned to the double grain-size %"G_GINT64_MODIFIER "d",
+				region_size, 2 * grain_size);
+		goto out;
+	}
+
+	number_of_sectors = fdisk_get_nsectors(cxt);
+	if ((region_start + region_size) / sector_size >= number_of_sectors) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Region configuration is bigger than device");
+		res = FALSE;
+		goto out;
+	}
+
+	sector_size = fdisk_get_sector_size(cxt);
+	region_start_sector = region_start / sector_size;
+	region_end_sector = region_start_sector + region_size / sector_size - 1;
+
+	if (fdisk_get_partitions(cxt, &tb)) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Failed to get partitions");
+		return FALSE;
```

Comment:
To stay consistent presumably this should be:
```suggestion
		goto out;
```